IndexedDbGetAllEntries/explainer.md

```diff
@@ -7,23 +7,38 @@
 
 For batched record iteration, for example, retrieving *N* records at a time, the primary and index keys provided by `getAllRecords()` can eliminate the need for an [`IDBCursor`](https://developer.mozilla.org/en-US/docs/Web/API/IDBCursor), which further reduces the number of JavaScript events required. To read the next *N* records, instead of advancing a cursor to determine the range of the next batch, getAllRecords() can use the primary key or the index key retrieved by the results from the previous batch.
 
+Update the existing operations `getAll()` and `getAllKeys()` to support the same query options as `getAllRecords()`, which adds direction.  For some scenarios, `getAll()` and `getAllKeys()` may suffice.  For example, developers may use `getAllKeys()` to defer loading values until needed.  For records with inline keys, `getAll()` already retrieves both key and value.
+
 ## `IDBObject::getAllRecords()` and `IDBIndex::getAllRecords()`
 
 This explainer proposes adding `getAllRecords()` to both [`IDBObjectStore`](https://www.w3.org/TR/IndexedDB/#idbobjectstore) and [`IDBIndex`](https://www.w3.org/TR/IndexedDB/#idbindex). `getAllRecords()` creates a new `IDBRequest` that queries its `IDBObjectStore` or `IDBIndex` owner. The `IDBRequest` completes with an array of `IDBRecord` results. Each `IDBRecord` contains the `key`, `primaryKey` and `value` attributes. For `IDBIndex`, `key` is the record's index key. For `IDBObjectStore`, both `key` and `primaryKey` return the same value. The pre-existing [`IDBCursorWithValue`](https://www.w3.org/TR/IndexedDB/#idbcursorwithvalue) interface contains the same attributes and values for both `IDBObjectStore` and `IDBIndex`. However, unlike `getAllRecords()`, a cursor may only read one record at a time.
 
+## Adding direction to `getAll()` and `getAllKeys()`
+
+This explainer proposes using `getAllRecords()` as feature detection for direction support in `getAllKeys()` and `getAll()`.  `getAllRecords()` introduces the `IDBGetAllOptions` dictionary, which developers may also use with `getAll()` and `getAllKeys()`.  Before using `IDBGetAllOptions`, developers must check for the existence of `getAllRecords()` in `IDBObjectStore` or `IDBIndex`.
+
+## Compatibility risk
+
+Overloading `getAll()` and `getAllKeys()` to accept the `IDBGetAllOptions` dictionary introduces compatibility risk.  Prior to this proposal, when passed a dictionary argument, both `getAll()` and `getAllKeys()` throw an exception after [failing to convert the dictionary to a key range](https://w3c.github.io/IndexedDB/#convert-a-value-to-a-key-range).  After the overload, `getAllKeys()` and `getAll()` will no longer throw for dictionary input.  When the `IDBGetAllOptions` dictionary initializes with its default values, it creates a query that retrieves all of the keys or values from the entire database.
+
+Since using a dictionary with `getAll()` and `getAllKeys()` is a programming error, we believe compat risk is low.
```

Comment:
We could potentially decrease the compat risk by making one of the dictionary members required like 'direction'.  Then we could still throw on any arbitrary dictionary input that maybe existed before this change.

Reply:
Seems undesirable since direction would otherwise be optional. Maybe something to call attention to during blink-dev review and address if someone pushes back?

Reply:
Yes, I agree making direction required seems unnatural and we should only pursue if absolutely necessary.